React Pages Migration: /users

qreoct opened the pull request:
Part of #4626 

Move the course users (course/XX/users) and course user profile page (course/XX/users/XX) to React.

## To do
- [x] Initial migration
- [x] Write tests
- [x] Remove unused Ruby code
- [x] Integrate newly built React components into Achievements refactor

## Course Users page
**Before**
![image](https://user-images.githubusercontent.com/9080974/171637054-bad211e3-5be6-4bcd-bc33-38298b44ae35.png)

**After**
Fixed alignment with long names
![image](https://user-images.githubusercontent.com/9080974/171634097-1a21bf14-dff6-4613-aed8-7f04cdd5392a.png)

## User Profile page
**Before**
![image](https://user-images.githubusercontent.com/9080974/171636765-3f43f8d5-9f15-4bd5-898c-cf2648eb0920.png)
(Classmate without achievements)
![image](https://user-images.githubusercontent.com/9080974/171636871-cfa12928-f6b2-48c4-8bc0-e93d95bd8721.png)
(Classmate view)
![image](https://user-images.githubusercontent.com/9080974/171636912-a6da4cdd-0943-49d6-97bb-f151781e2ea1.png)


**After**
(Viewing own profile) -- Can manage email subscriptions & view EXP history (click on EXP card).
![user profile page](https://user-images.githubusercontent.com/9080974/171634540-31ee9cd4-2841-4b20-8aed-c0940c8dc321.png)

(Viewing classmate profile) -- Only able to see achievements, shouldn't see exp/level (same as current behaviour)
![image](https://user-images.githubusercontent.com/9080974/171634921-3f3551f4-f7ea-4767-888a-c2636bd26f89.png)
(Classmate without achievements -- shows empty state)
![image](https://user-images.githubusercontent.com/9080974/171634770-b8b8f6f9-a399-4c52-a480-6a418c10dbbd.png)

(Viewing teaching staff profile) -- Shouldn't see gamified elements
![image](https://user-images.githubusercontent.com/9080974/171634662-b5adbae6-5f03-4927-bb38-97af9308a43e.png)

(Achivements Component disabled)
![image](https://user-images.githubusercontent.com/9080974/171635414-f2a37bc8-c8f2-4f5d-87eb-91baeb7c757b.png)

(Gamification elements disabled)
![image](https://user-images.githubusercontent.com/9080974/171635330-1361e2f3-b603-4d26-b93e-d5dc6a3f3a83.png)

ekowidianto commented on client/app/types/course/users.ts:
```diff
@@ -0,0 +1,48 @@
+import { Permissions } from 'types';
```
there is course_users.ts

Could we move everything here to that file and call it CourseUserXXX instead? The reason for this is because we need to differentiate courseUser and User in the future.

ekowidianto commented on client/app/bundles/course/users/types.ts:
```diff
@@ -0,0 +1,33 @@
+// State Types
+
+export interface UsersState {
+  permissions: UserPermissions;
```
Remove if not needed

ekowidianto commented on client/app/bundles/course/users/types.ts:
```diff
@@ -0,0 +1,33 @@
+
+// Action Names
+export const SAVE_USER = 'course/users/SAVE_USER';
+export const SAVE_USERS_LIST = 'course/users/SAVE_USERS_LIST';
```
Put this before SAVE_USER for consistency :)

ekowidianto commented on client/app/bundles/course/users/types.ts:
```diff
@@ -0,0 +1,33 @@
+  users: EntityStore<UserMiniEntity, UserEntity>;
+}
+
+export type UsersActionType = SaveUsersListAction | SaveUserAction;
```
Can we move this above the // State Types comment?

ekowidianto commented on client/app/bundles/course/users/reducers.ts:
```diff
@@ -0,0 +1,43 @@
+        ...data,
+      }));
+      saveListToStore(draft.users, entityList);
+      draft.permissions = action.userPermissions;
```
Again, remove if not needed.

ekowidianto commented on client/app/bundles/course/users/operations.ts:
```diff
@@ -0,0 +1,25 @@
+      .index()
+      .then((response) => {
+        const data = response.data;
+        dispatch(actions.saveUsersList(data.users, data.permissions));
```
Remove permissions if not needed

ekowidianto commented on client/app/bundles/course/users/actions.ts:
```diff
@@ -0,0 +1,25 @@
+  };
+}
+
+export function saveUsersList(
```
Lets put this above saveUser

ekowidianto commented on client/app/bundles/course/users/pages/UsersIndex/index.tsx:
```diff
@@ -0,0 +1,103 @@
+          defaultMessage: 'Students',
+        })}
+      />
+      <Grid container>
```
You can use <Note /> component if courseUsers is empty

ekowidianto commented on client/app/bundles/course/users/pages/UserShow/index.tsx:
```diff
@@ -0,0 +1,67 @@
+    return null;
+  }
+
+  const styles = {
```
Can we move this out from the UserShow component?

ekowidianto commented on client/app/bundles/course/users/pages/UserShow/index.tsx:
```diff
@@ -0,0 +1,67 @@
+
+  return (
+    <Box sx={styles.userShowPage}>
+      <UserProfileCard
```
Rather than passing 1 by 1, it would be better to pass the whole user object and destructure it inside the component.

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+  manageEmailSubscriptionUrl,
+  achievementCount,
+}: Props) => {
+  const styles = {
```
Let's move styles outside of the main component

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+    },
+  };
+
+  function UserStatsCard(props): JSX.Element {
```
Why not use const here?

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+    className: PropTypes.string,
+  };
+
+  function handleScrollToAchievements(e: React.MouseEvent): void {
```
Same here

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+    );
+  }
+
+  UserStatsCard.propTypes = {
```
Why do we have propTypes here? I suppose we are already using TS and thus, we dont need this anymore

ekowidianto replied:
You can declare a local interface instead for TS

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+        </Box>
+      );
+    }
+    return <>{email}</>;
```
I think we can just do `return email`

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+            variant="body1"
+            underline="hover"
+          >
+            Manage Email Subscriptions
```
Use react-intl

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+        justifyContent={{ xs: 'center', sm: 'start' }}
+        sx={styles.statsContainer}
+      >
+        {level !== undefined && (
```
Why not use `level && (` ? Same for the rest below

qreoct replied:
Level (and exp) can be 0, which is a falsy value and will not render the component

ekowidianto replied:
Thanks for clarifying!

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+import { scroller } from 'react-scroll';
+import { blue, green, orange } from '@mui/material/colors';
+
+interface Props {
```
So if you pass the user instead, you can use UserEntity as the type of the props instead of re-declaring the interface

ekowidianto commented on client/app/bundles/course/users/components/UserProfileAchievements.tsx:
```diff
@@ -0,0 +1,78 @@
+});
+
+const UserProfileAchievements: FC<Props> = ({ achievements, intl }: Props) => {
+  const styles = {
```
Similarly, move it out

ekowidianto commented on client/app/bundles/course/users/components/UserProfileAchievements.tsx:
```diff
@@ -0,0 +1,78 @@
+  return (
+    <>
+      <Typography variant="h4" component="h2" id="user-profile-achievements">
+        Achievements
```
use react-intl

ekowidianto commented on client/app/bundles/course/users/components/UserProfileSkills.tsx:
```diff
@@ -0,0 +1,59 @@
+      <Typography variant="h4" component="h2">
+        Topic Mastery
+      </Typography>
+      <Table>
```
Got it!

ekowidianto commented on client/app/bundles/course/users/components/UserProfileSkills.tsx:
```diff
@@ -0,0 +1,59 @@
+      </Typography>
+      <Table>
+        <TableBody>
+          {skillBranches.map((skillBranch) => (
```
Lets handle the case where skillBranches is empty

ekowidianto commented on client/app/bundles/course/users/components/UserProfileSkills.tsx:
```diff
@@ -0,0 +1,59 @@
+import React, { FC } from 'react';
```
`import { FC, Fragment } from 'react';`

Lets used named import

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,193 @@
+        maxWidth: '144px',
+        transition: 'background-color 0.5s, color 0.5s',
+      },
+      '& :nth-child(2n):hover .user-stats-card,& .user-stats-card:nth-child(2n):hover':
```
Ive got this error for using nth-child.

![image](https://user-images.githubusercontent.com/42570513/171787914-b943dfc8-f587-434d-abed-ef43480fa272.png)

qreoct replied:
Shifted to course/users.scss since we need nth-child for this particular styling. A bit overkill to use emotion's `emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason` flag, even though we are not doing server-side rendering.

ekowidianto reviewed (changes requested):
Good work @qreoct for your first react page migration ! I like the new layout! Some comments and inputs below.

ekowidianto commented:
Please also check if we need these 2 files:
app/views/course/users/_course_user.html.slim
app/views/course/users/_achievement.html.slim

ekowidianto commented on client/app/bundles/course/users/components/UserProfileAchievements.tsx:
```diff
@@ -0,0 +1,78 @@
+            >
+              <Grid container sx={styles.achievementMiniEntityContainer}>
+                <a
+                  href={`/courses/${getCourseId()}/achievements/${
```
Can abstract this and create a new helper function in url-builders.js

qreoct commented on app/views/course/users/_user_data.json.jbuilder:
```diff
@@ -0,0 +1,47 @@
+all_skill_branches = @skills_service.skill_branches
+can_view_skills = all_skill_branches.present? && can_read_progress
+if can_view_skills && is_student_and_gamified
+  json.skillBranches all_skill_branches.each do |skill_branch|
```
Good idea!

qreoct commented on spec/features/course/user_profile_spec.rb:
```diff
@@ -4,44 +4,42 @@
 RSpec.feature 'Courses: CourseUser Profile' do
   let(:instance) { Instance.default }
 
-  def exp_text(course_user)
```
Added tests for this!

qreoct commented:
Thanks Eko for the code review! I've addressed the comments, strengthened the tests, and added empty states for certain components:
Students Index Page
![image](https://user-images.githubusercontent.com/9080974/171831066-d733fe3d-a003-4c78-a85f-f2d45a7b4e42.png)

Skill Branches
![image](https://user-images.githubusercontent.com/9080974/171831089-1a7168bb-2c08-4ff1-a731-3f6008f0696a.png)

ekowidianto commented on app/assets/stylesheets/course/users.scss:
```diff
@@ -12,47 +12,15 @@
     }
   }
-
   &.show {
```
Just create a scss file in react frontend and remove everything in &.show here

ekowidianto commented on app/views/course/users/_user_data.json.jbuilder:
```diff
@@ -0,0 +1,53 @@
+if can_view_skills && is_student_and_gamified
+  json.partial! 'course/assessment/skill_branches/skill_branch', skill_branches: all_skill_branches
+
+  # json.skillBranches all_skill_branches.each do |skill_branch|
```
Remove unused comments

ekowidianto commented on client/app/bundles/course/users/components/UserProfileCard.tsx:
```diff
@@ -0,0 +1,198 @@
+});
+
+const UserProfileCard: FC<Props> = ({ user, intl }) => {
+  interface UserStatsCardProps {
```
Can move this outside the functional component.

ekowidianto reviewed (approved):
Thanks for updating! Awesome response time! :) Some minor nits but it seems like we are good to go!

Also please include the achievement course user profile picture list refactor in another commit in this PR.

cysjonathan commented:
@qreoct can consider using the same status bar as the assessment submissions page for the Topic mastery section.
percentage % can leave outside on the right of the bar.

Pros:
1. Can indicate number of "topics" uncompleted / completed within the progress bar itself, which makes use of the wide space occupied by the horizontal progress bar.
2. Can recycle a common component + colour palette, which helps keep a sense of consistency for users, and reduces the number of different overall components we manage.

Cons:
1. need to rewrite

Up for discussion. If this is useful/meaningful, we can save these for a separate PR.
